### PR TITLE
Ipns/patches

### PR DESCRIPTION
**Pull request by whyrusleeping**
This PR contains three distinct commits that make ipns more reliable as a temporary fix until finalize the ipld records implementation.

Details on each commit provided in commit messages.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
weird failures i'm fairly certain arent related: https://travis-ci.org/ipfs/go-ipfs/jobs/81498938#L247

File: core/core.go [hunk 1]
```diff
@@ -104,6 +105,7 @@ type IpfsNode struct {
 	Diagnostics  *diag.Diagnostics   // the diagnostics service
 	Ping         *ping.PingService
 	Reprovider   *rp.Reprovider // the value reprovider system
+	IpnsRepub    *ipnsrp.Republisher
```
Comment (jbenet) [inline core/core.go]: We'll eventually want to organize this stuff cleaner. maybe bundle all the IPNS parts into one struct.

File: core/core.go [hunk 2]
```diff
@@ -226,6 +228,12 @@ func (n *IpfsNode) startOnlineServicesWithHost(ctx context.Context, host p2phost
 	// setup name system
 	n.Namesys = namesys.NewNameSystem(n.Routing)
 
+	// setup ipns republishing
+	n.IpnsRepub = ipnsrp.NewRepublisher(n.Routing, n.Peerstore)
+	n.IpnsRepub.AddName(n.Identity)
+
+	n.Process().Go(n.IpnsRepub.Run)
```
Comment (jbenet) [inline core/core.go]: idk, maybe `NewRepublisher` should launch this, or have:
```go
n.IpnsRepub.Run(n.Process()) // calls proc.Go(func) inside.
```

seems like too low level ctrl

File: namesys/publisher.go [hunk 3]
```diff
@@ -110,6 +124,30 @@ var IpnsRecordValidator = &record.ValidChecker{
 	Sign: true,
 }
 
+func IpnsSelectorFunc(k key.Key, vals [][]byte) (int, error) {
+	var recs []*pb.IpnsEntry
+	for _, v := range vals {
+		e := new(pb.IpnsEntry)
+		err := proto.Unmarshal(v, e)
+		if err != nil {
+			// Probably dont need to fully error out, other records might be valid
```
Comment (jbenet) [inline namesys/publisher.go]: yep, continue.

Comment (jbenet) [inline namesys/publisher.go]: This function should just sort them and return them sorted. see the records spec for why.

File: namesys/publisher.go [hunk 4]
```diff
@@ -110,6 +124,30 @@ var IpnsRecordValidator = &record.ValidChecker{
...
+	var best_i int
+
+	for i, r := range recs {
+		if r.GetSequence() > best_seq {
+			best_seq = r.GetSequence()
+			best_i = i
+		}
+	}
+
+	return best_i, nil
```
Comment (jbenet) [inline namesys/publisher.go]: This function should just sort them and return them sorted. see the records spec for why.

Comment (jbenet) [inline namesys/publisher.go]: (then pick the first)

File: namesys/republisher/repub.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		val := vals[0].Val
+		e := new(pb.IpnsEntry)
+		err = proto.Unmarshal(val, e)
+		if err != nil {
+			return err
+		}
+		p := path.Path(e.Value)
+
+		// republish it
+		err = rp.pub.Publish(ctx, priv, p)
```
Comment (jbenet) [inline namesys/republisher/repub.go]: i believe this will increment the seqno. not convinced what we gain by incrementing seqno when value is the same.

**Comment by jbenet** [comment]
needs some more tests. can manipulate the timeouts to test the various edge cases.

File: core/core.go (hunk 1 above)
Comment (whyrusleeping) [inline core/core.go]: That makes sense. an `IpfsNode` does a *lot*

File: namesys/republisher/repub.go (hunk 5 above)
Comment (whyrusleeping) [inline namesys/republisher/repub.go]: yeah, we discussed that on hangouts. This is what you said was the right way to do it.

File: namesys/publisher.go (hunk 4 above)
Comment (whyrusleeping) [inline namesys/publisher.go]: > see the records spec for why.

went through the records spec, is says *to* order them, but not why.

File: namesys/republisher/repub.go (hunk 5 above)
Comment (jbenet) [inline namesys/republisher/repub.go]: i didn't mean increment the seqno on every republish, i meant increment the seqno every time the value is changed. e.g.:

```
ipfs name publish A # increments seqno
ipfs name publish B # increments seqno
wait 10000000 # republishes happen without incrementing seqno
```

Comment (jbenet) [inline namesys/republisher/repub.go]: (is there a reason you need to incr the seqno on republish?)

File: namesys/publisher.go (hunk 4 above)
Comment (jbenet) [inline namesys/publisher.go]: the spec calls for having a total order. (a total order over records means that you can _always_ sort and thus pick between any given number of records, deterministically.)

File: namesys/republisher/repub.go (hunk 5 above)
Comment (whyrusleeping) [inline namesys/republisher/repub.go]: > i didn't mean increment the seqno on every republish, i meant increment the seqno every time the value is changed.

gotcha, must have misunderstood you in chat.

We dont need to increment it, doing so just makes the record sorting algorithm that much simpler.

File: core/core.go (hunk 2 above)
Comment (whyrusleeping) [inline core/core.go]: Then my function starts to look pretty messy:
```
func (rp *Republisher) Run(p gp.Process) {
    p.Go(func(proc gp.Process) {
        // code 
    }
}
```

which to me seems kinda wonky. My intent is to essentially do (with goprocess) `go n.IpnsRepub.Run(ctx)` which has a fairly simple and readable intent.

Comment (jbenet) [inline core/core.go]: ok :+1:

File: namesys/republisher/repub.go (hunk 5 above)
Comment (jbenet) [inline namesys/republisher/repub.go]: easier to not have to write things again to storage if it's the same record. storage may be S3.

Comment (whyrusleeping) [inline namesys/republisher/repub.go]: we will have to write it again anyways since we're updating the timestamp

Comment (jbenet) [inline namesys/republisher/repub.go]: ah yes, that's true. regardless, the generation number growing at every republish isn't a good thing. WAAAAAY easier to reason about things if an increment means and intended, logical change.

Comment (whyrusleeping) [inline namesys/republisher/repub.go]: yeap! agreed.

**Comment by whyrusleeping** [comment]
@jbenet what tests are you thinking of? About all I could come up with is to shorten the interval and check that the publish actually happened.

**Comment by jbenet** [comment]
- check the publish happens
- check republish happens
- check records disappear after interval if (re)publisher goes down.
- check records re-appear if publisher returns.

(what was the problem with iptb on travis?)

**Comment by whyrusleeping** [comment]
the issue with iptb was that it didnt use 'port zero' since it would have had to start up a daemon, get the address it chose from the output, and write that to the other nodes bootstrap configs. I still use iptb all the time locally

**Comment by whyrusleeping** [comment]
hey look! our favorite type of race/panic!

https://travis-ci.org/ipfs/go-ipfs/jobs/82099648

**Comment by whyrusleeping** [comment]
@jbenet I've added a test for the republisher

**Comment by jbenet** [comment]
@whyrusleeping durability concerns o/

**Comment by whyrusleeping** [comment]
@jbenet ping. There isnt anything to worry about here as far as durability goes. It the node goes down, we restart it and things continue where they left off.

File: core/core.go [hunk 6]
```diff
@@ -226,6 +228,12 @@ func (n *IpfsNode) startOnlineServicesWithHost(ctx context.Context, host p2phost
 	// setup name system
 	n.Namesys = namesys.NewNameSystem(n.Routing)
 
+	// setup ipns republishing
+	n.IpnsRepub = ipnsrp.NewRepublisher(n.Routing, n.Repo.Datastore(), n.Peerstore)
+	n.IpnsRepub.AddName(n.Identity)
```
Comment (jbenet) [inline core/core.go]: ok so this is why we don't yet need durability o/ -- we provide durability elsewhere, and always load up the state ourselves.

File: namesys/publisher.go [hunk 7]
```diff
@@ -44,56 +46,102 @@ func NewRoutingPublisher(route routing.IpfsRouting) Publisher {
 // and publishes it out to the routing system
 func (p *ipnsPublisher) Publish(ctx context.Context, k ci.PrivKey, value path.Path) error {
 	log.Debugf("Publish %s", value)
+	return p.PublishWithEOL(ctx, k, value, time.Now().Add(time.Hour*24))
```
Comment (jbenet) [inline namesys/publisher.go]: maybe add a note for people watching/reading the code that this is a step in between today, and https://github.com/ipfs/specs/tree/master/records

Comment (jbenet) [inline namesys/publisher.go]: (a lot of people have expectations about how ipns and records in general will work for their app)

File: namesys/publisher.go [hunk 8]
```diff
@@ -44,56 +46,102 @@ func NewRoutingPublisher(route routing.IpfsRouting) Publisher {
...
 	if err != nil {
 		return err
 	}
 
-	ipnskey := key.Key("/ipns/" + string(nameb))
+	return nil
+}
+
+func PublishEntry(ctx context.Context, r routing.IpfsRouting, ipnskey key.Key, rec *pb.IpnsEntry) error {
+	timectx, cancel := context.WithDeadline(ctx, time.Now().Add(time.Second*10))
```
Comment (jbenet) [inline namesys/publisher.go]: 10s seems small. 30s/60s maybe. some networks suck.

Comment (jbenet) [inline namesys/publisher.go]: oh, actually this is a DHT query, 60s more like it.

File: namesys/publisher.go (hunk 7 above)
Comment (whyrusleeping) [inline namesys/publisher.go]: ah, right. will do

File: namesys/publisher.go (hunk 8 above)
Comment (whyrusleeping) [inline namesys/publisher.go]: i definitely didnt notice that was only ten seconds... good catch.

File: namesys/republisher/repub_test.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	// Now wait a second, the records will be invalid and we should fail to resolve
+	time.Sleep(time.Second)
+	if err := verifyResolutionFails(nodes, name); err != nil {
+		t.Fatal(err)
+	}
+
+	// The republishers that are contained within the nodes have their timeout set
+	// to 12 hours. Instead of trying to tweak those, we're just going to pretend
+	// they dont exist and make our own.
```
Comment (jbenet) [inline namesys/republisher/repub_test.go]: yeah, its good to test both: 
- that republishers work
- that republishers installed on the node work

the more tests the better, as it is we need way more coverage of all our expectations.

fine to merge for now, but we'll need something like this eventually. will be easier when we split things out

**Comment by jbenet** [comment]
Feeling good about this. i think we should have some sharness and/or iptb tests around this. it's a large complicated system with lots of expectations.

File: namesys/republisher/repub_test.go (hunk 9 above)
Comment (whyrusleeping) [inline namesys/republisher/repub_test.go]: yeah, muuuuuch easier once we get to the real implementation.

**Comment by whyrusleeping** [comment]
@jbenet So, i fixed iptb so that it should work with port zero stuff. and then i realized that in order to be able to tweak the timeouts and intervals of the rebroadcaster through the cli, it would take a lot of reworking things... do you think thats worth the time investment?

**Comment by jbenet** [comment]
Could be a config option. Probably should be so we can play with it. 




May also want to make it default to every 4hrs for now actually, because network is still small and churn. Prob confirm that it isn't less than every 5min



—

**Comment by whyrusleeping** [comment]
@jbenet what would we call this config option? I have the change to four hours in, and it shouldnt be difficult at all to grab a value from the config if it exists, just gotta decide on where it lives.

**Comment by jbenet** [comment]
`IPNS.RepublishPeriod` and a `time.Duration` string?

**Comment by whyrusleeping** [comment]
alright, we've got a config setting!

**Comment by whyrusleeping** [comment]
@jbenet next steps?

**Comment by jbenet** [comment]
@whyrusleeping: 
- https://github.com/ipfs/go-ipfs/pull/1739#commitcomment-13455202
- https://github.com/ipfs/go-ipfs/pull/1739#commitcomment-13455207

**Comment by whyrusleeping** [comment]
@jbenet error cases added.

**Comment by jbenet** [comment]
@whyrusleeping tests on the config value. there is a config sharness test for this reason.

I still think this is not at all tested enough, and am not comfortable with it. Now that iptb is fixed and working with sharness, we should be able to test republishing happens correctly. I'm ok to merge it since it seems strictly better than what we have now, but this _really, really should be well tested_. it's a very important part of the system.

**Comment by jbenet** [comment]
> in order to be able to tweak the timeouts and intervals of the rebroadcaster through the cli

these probably should be configurable anyway, and if not, use ENV vars that only work if `DEBUG` is true

**Comment by whyrusleeping** [comment]
I'm gonna have to make it not error out if we pass a smaller than one minute value for the republish timeout. Otherwise the tests will actually take multiple whole minutes to run

**Comment by jbenet** [comment]
@whyrusleeping not really, one possibility; have that check normally AND allow a smaller retransmit IF `DEBUG` is true (as mentioned above). basically:

```
ipfs daemon # errors out if value to small
ipfs daemon --debug # does not error out if value too small
```
that way can use `--debug` for the tests. (IPFS_TESTING=true or some other env).

**Comment by whyrusleeping** [comment]
Things are looking as good as i think theyre gonna.  The sharness test seems to fail occasionally. I'm guessing something takes a long time for no reason at all and the republisher fires. ping @jbenet

**Comment by jbenet** [comment]
@whyrusleeping feedback above o/

**Comment by whyrusleeping** [comment]
I think it hangs because of a potential bug in the dht query code. On a localhost connection like this, it should quickly exhaust all nodes in the network in a query, and stop the query when it gets there...

**Comment by whyrusleeping** [comment]
@jbenet alright, lookin good!

**Comment by whyrusleeping** [comment]
aaaand rebased.

**Comment by whyrusleeping** [comment]
@jbenet RFM

**Comment by jbenet** [comment]
The diff is too big to comment inline :/
- [ ] been discussing for months that this needs to be total order https://github.com/ipfs/go-ipfs/commit/cfcc3d6a1b77a47cdf6512e6be24e8b3bbd9fb05#commitcomment-13512298
- [ ] get more records, bound with a timeout, exhaust https://github.com/ipfs/go-ipfs/commit/cfcc3d6a1b77a47cdf6512e6be24e8b3bbd9fb05#commitcomment-13512408
- [ ] ctx cancels https://github.com/ipfs/go-ipfs/commit/cfcc3d6a1b77a47cdf6512e6be24e8b3bbd9fb05#commitcomment-13512370
- [ ] chan? (meh) https://github.com/ipfs/go-ipfs/commit/cfcc3d6a1b77a47cdf6512e6be24e8b3bbd9fb05#commitcomment-13512439
- [ ] rateLimit not enough? https://github.com/ipfs/go-ipfs/commit/235a9ec5fcbb2ba6cc9099e3d77ef386ab62398b#commitcomment-13512520
- [ ] where decr? https://github.com/ipfs/go-ipfs/commit/235a9ec5fcbb2ba6cc9099e3d77ef386ab62398b#commitcomment-13512529
- [ ] publish key note (fine not to do in this PR) https://github.com/ipfs/go-ipfs/commit/815a031f595bbe68d084c2666e0c353469bcb05a#commitcomment-13512557
- [ ] nicer err for user https://github.com/ipfs/go-ipfs/commit/815a031f595bbe68d084c2666e0c353469bcb05a#commitcomment-13512589
- [ ] dedup important publishing code https://github.com/ipfs/go-ipfs/commit/815a031f595bbe68d084c2666e0c353469bcb05a#commitcomment-13512668
- [ ] can you split this commit in two? i cant CR it -- diff too big: https://github.com/ipfs/go-ipfs/commit/41db338751ef8b04e1ebe7a9f6a0767df38e148e maybe vendor deps first, then the changes)

(what is `kingpin` used for?)

**Comment by jbenet** [comment]
@whyrusleeping

**Comment by whyrusleeping** [comment]
kingpin is the flags library that iptb uses.

**Comment by whyrusleeping** [comment]
@jbenet addressed comments, and split up the vendoring commit.

**Comment by jbenet** [comment]
- [ ] total ordering https://github.com/ipfs/go-ipfs/commit/3c7c1657859186a9204e2f1daa6bcfa938ee55d0#commitcomment-13527896
- [ ] context cancel fix https://github.com/ipfs/go-ipfs/commit/3c7c1657859186a9204e2f1daa6bcfa938ee55d0#commitcomment-13527951

**Comment by whyrusleeping** [comment]
did those things, @jbenet the things were done did.

**Comment by jbenet** [comment]
@whyrusleeping great! thanks! :clap: :clap: 

I'm not 100% this will work perfectly, but let's merge it and see how it does?
